[RDY] Implement persistent history.

Balletie opened the pull request:
I added an `IInputHistory` interface, so that we have an interface to talk to with other parts of our repl. I then created an adapter for JLine2's history implementation.

Balletie committed:
Set the History to a FileHistory for persistency

Balletie committed:
Add IInputHistory interface, switch to adapter for jline2 history.

Balletie committed:
Add tests for both Adapters (persistent and non-persistent)

Hjdskes commented on org.metaborg.spoofax.shell.console/src/main/java/org/metaborg/spoofax/shell/client/console/ConsoleReplModule.java:
```diff
@@ -48,15 +51,15 @@ protected void configure() {
-                                                 @Named("out") OutputStream out)
-                                                     throws IOException {
-        return new ConsoleReader(in, out);
+    protected jline.console.ConsoleReader provideConsoleReader(@Named("in") InputStream in,
```
Why not import the package?

Balletie replied:
This one can be imported. I did not want to import jline's History classes so that the difference between the adapter and the "adapted" is clear.

Balletie replied:
Then it's clear that it's a library, not from us. But I can turn it back if you want.

Hjdskes replied:
@gfokkema thoughts? I prefer to import things unless we can't.

Hjdskes commented on org.metaborg.spoofax.shell.console/src/main/java/org/metaborg/spoofax/shell/client/console/JLine2InputHistory.java:
```diff
@@ -0,0 +1,68 @@
+ * An adapter for JLine2's History implementation.
+ */
+public class JLine2InputHistory implements IInputHistory {
+    protected jline.console.ConsoleReader reader;
```
Why not import the package?

Hjdskes commented on org.metaborg.spoofax.shell.console/src/main/java/org/metaborg/spoofax/shell/client/console/JLine2PersistentInputHistory.java:
```diff
@@ -0,0 +1,80 @@
+        jline.console.history.FileHistory newDelegate =
+            new jline.console.history.FileHistory(histFile);
+
+        // Have to clear so that I can add all the old entries in front again...
```
Does this have to be implemented?

Balletie replied:
I can remove it, when you load from disk it will remove the history that was in memory.

Hjdskes replied:
I meant; should this be implemented before we merge this PR? The functionality (loading disk and inserting that before what is currently in memory) is important imo.

Balletie replied:
~~Yea, it's just that it's rather hacky. There's no other workaround because their implementation is a bit stupid :p.~~

Actually, now that I think of it, the in-memory contents should be added AFTER loading the history from file, because the in-memory entries are the most recent. So there's no workaround necessary at all :P.

Balletie replied:
Fixed it in 9a5c5fb.

Hjdskes commented on org.metaborg.spoofax.shell.console/src/main/java/org/metaborg/spoofax/shell/client/console/JLine2InputHistory.java:
```diff
@@ -0,0 +1,68 @@
+ */
+public class JLine2InputHistory implements IInputHistory {
+    protected jline.console.ConsoleReader reader;
+    private jline.console.history.MemoryHistory delegateHist;
```
`final`?

Hjdskes commented on org.metaborg.spoofax.shell.console/src/main/java/org/metaborg/spoofax/shell/client/console/JLine2InputHistory.java:
```diff
@@ -0,0 +1,68 @@
+     *            The jline2 history implementation that will be delegated to.
+     */
+    @Inject
+    JLine2InputHistory(jline.console.ConsoleReader reader,
```
`final` for parameters?

Hjdskes commented on org.metaborg.spoofax.shell.console/src/main/java/org/metaborg/spoofax/shell/client/console/JLine2PersistentInputHistory.java:
```diff
@@ -0,0 +1,80 @@
+ * again using {@link #persistToDisk()}.
+ */
+public class JLine2PersistentInputHistory extends JLine2InputHistory {
+    private String filePath;
```
`final`?

Hjdskes commented on org.metaborg.spoofax.shell.console/src/main/java/org/metaborg/spoofax/shell/client/console/JLine2PersistentInputHistory.java:
```diff
@@ -0,0 +1,80 @@
+
+    @Override
+    public void loadFromDisk() throws IOException {
+        File histFile = new File(filePath);
```
`final`?

Hjdskes commented on org.metaborg.spoofax.shell.console/src/test/java/org/metaborg/spoofax/shell/client/console/TerminalUserInterfaceTest.java:
```diff
@@ -24,8 +24,8 @@
     private TerminalUserInterface ui;
-    private static final String PROMPT = "<TEST>";
-    private static final String CONT_PROMPT = ".TEST.";
+    public static final String PROMPT = "<TEST>";
```
Maybe create a new class to host this kind of things

Balletie replied:
Meh, it's good enough for a test.

Balletie committed:
Use final where possible

Balletie committed:
Add the in-memory entries after switching to persistent history

The in-memory entries are the most recent, so they're added after all of
the loaded entries have been added.

Balletie commented on org.metaborg.spoofax.shell.console/src/main/java/org/metaborg/spoofax/shell/client/console/JLine2InputHistory.java:
```diff
@@ -0,0 +1,68 @@
+ */
+public class JLine2InputHistory implements IInputHistory {
+    protected jline.console.ConsoleReader reader;
+    private final jline.console.history.MemoryHistory delegateHist;
```
You want me to import it here too?

Hjdskes replied:
If possible :)

Balletie committed:
Fix JavaDoc errors in one of the tests
